### PR TITLE
[SPARK-21402][SQL][BACKPORT-2.2] Fix java array of structs deserialization

### PR DESCRIPTION
**Pull request by vofque**
This PR is to backport #22708 to branch 2.2.

## What changes were proposed in this pull request?

MapObjects expression is used to map array elements to java beans. Struct type of elements is inferred from java bean structure and ends up with mixed up field order.
I used UnresolvedMapObjects instead of MapObjects, which allows to provide element type for MapObjects during analysis based on the resolved input data, not on the java bean.

## How was this patch tested?

Added a test case.
Built complete project on travis.

@dongjoon-hyun @cloud-fan

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanWithArraySuite.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      .format("json")
+      .schema(createSchema())
+      .load("src/test/resources/test-data/with-array-fields.json")
+      .as(encoder);
+
+    List<Record> records = dataset.collectAsList();
+    Assert.assertEquals(records, RECORDS);
+  }
+
+  private StructType createSchema() {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanWithArraySuite.java]: So, this is back inevitably. :)

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanWithArraySuite.java (hunk 1 above)
Comment (vofque) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanWithArraySuite.java]: Yeah, it managed to survive

**Comment by dongjoon-hyun** [comment]
Merged to `branch-2.2`.

**Comment by dongjoon-hyun** [comment]
Thank you, @vofque .

**Comment by dongjoon-hyun** [comment]
Could you close this PR? Auto-close is only working for `master` branch PR.

**Pull request closed without merging**